Create new Read the docs theme with custom index

jessica-mitchell opened the pull request:
This PR introduces a new theme (based on Sphinx Material) for the Read the docs documentation. 
It also include a new template for index.html. Thanks to @jougs for developing a more interactive and attractive theme!

The goal is to provide users with different ways to explore the docs, improve the context for the different guides and resources, and enhance the overall theme.


Set as draft as discussion and bit more tweaking is needed. Also  #2283 needs to be merged first.

jessica-mitchell commented:
@heplesser @steffengraber This PR touches a lot of files, but most is fixing up minor layout problems and broken links, and adding the grid-card directive to navigation pages. A lot of customization was added to the CSS.  Part of the goals here is to simplify the navigation and present the documentation in more logical categories. 

The organization of the documentation topics hopefully makes sense from a content perspective, and from the presentation perspective, the HTML output should look correct and the layout shouldn't be broken in any major or strange way. I think the following pages show this new style best: 
documentation homepage: https://nest-test.readthedocs.io/en/rtd-theme/
install pages: https://nest-test.readthedocs.io/en/rtd-theme/installation/index.html
PyNEST examples: https://nest-test.readthedocs.io/en/rtd-theme/examples/index.html#pynest-examples
Understanding how NEST works: https://nest-test.readthedocs.io/en/rtd-theme/understand_index.html
NEST developer: https://nest-test.readthedocs.io/en/rtd-theme/developer_space/index.html

Might also be of interest to see discussions in #2283 regarding the restructing of the documentation.

heplesser commented:
@jessica-mitchell My first impression is that this is a welcome visual re-do of our documentation. But I also noticed a few problems. The core problem seems to be that the layout requires pretty wide browser windows to work properly (about 60% of screen width on a 30" screen). These are just first impressions.

- The font for "nest::docs" seems different from the font we use in the NEST logos
- The orange bar on top takes up a lot of screen real estate. Could we reduce the height?
- When I set the width of my browser window to a width I ususally use, the page looks strange, in particular, the navigation menu to the left disappears to somewhere to the left of the big orange bar, see screenshot. This happens with Safari and Firefox.
![Screenshot 2022-06-01 at 07 52 58](https://user-images.githubusercontent.com/1059644/171337549-710de7d4-10e9-43c8-9a1d-2bea7ee34837.png)
- I find the line breaks in the headlines of the tiles not pleasing, since the first line is next to the logo and subsequent lines below it. Is it possible to keep all header lines next to the logo or below?
![Screenshot 2022-06-01 at 07 58 08](https://user-images.githubusercontent.com/1059644/171337745-4ac6ae64-7862-4ff2-9a2b-9d9896a5dd5b.png)
- With my standard window width, the examples page looks not particularly good with four very narrow and tall columns. I think two columns would work better here.
- On mobile phone or tablet, the pages do not render well (Firefox).

jessica-mitchell commented:
> @jessica-mitchell My first impression is that this is a welcome visual re-do of our documentation. But I also noticed a few problems. The core problem seems to be that the layout requires pretty wide browser windows to work properly (about 60% of screen width on a 30" screen). These are just first impressions.
> 
>     * The font for "nest::docs" seems different from the font we use in the NEST logos
fixed
>     * The orange bar on top takes up a lot of screen real estate. Could we reduce the height?

this is the standard design of the sphinx-material theme, and I would prefer not to change theme for purely aesthetic reasons, at least for this PR.

>     * When I set the width of my browser window to a width I ususally use, the page looks strange, in particular, the navigation menu to the left disappears to somewhere to the left of the big orange bar, see screenshot. This happens with Safari and Firefox.

I think I fixed the strange problems for smaller screens and devices so it should be more readable and functional 
> 
>     * I find the line breaks in the headlines of the tiles not pleasing, since the first line is next to the logo and subsequent lines below it. Is it possible to keep all header lines next to the logo or below?

line breaks should be better - 
> 
>     * With my standard window width, the examples page looks not particularly good with four very narrow and tall columns. I think two columns would work better here.

The number of columns is changed based on screen size going from 1 to 3
> 
>     * On mobile phone or tablet, the pages do not render well (Firefox).

this should be fixed

jessica-mitchell commented:
@jougs after our discussion here are the changes I managed (also @heplesser @steffengraber fyi)
* css
    * Upgraded bootstrap from 3 to 5 - which resolved a lot issues (including bad font sizes and margins, and device scaling issues) that required some hacky css. However, I was unable to get the pulse button positioned and scaled properly so I removed it. I can look at re-adding at some point with fresh eyes
 
    * Removed unused css (erred on the side of keeping things if unsure)

    * Merged all custom css into one custom.css file and organized the sections to 1) basic / universal properties 2) material design 3) sphinx design 4) custom index.html 5) media queries (include all possible above categories)

    * Defined the nest-colors (orange, green and blue from index.html) as variables in css



* index.html
   * removed orange links on green banner
   * changed to row of buttons for links into docs (this seems ok)
   * font style is consistent with rest of documentation (with the caveat that the weight is heavier) 
   * fade particle image both borders 
   * fixed script highlighting

* specific formatting
    * pynest example pngs now all same size and line up 
    * changed how icons are placed in cards; this allows for easier control and consistency. icons are now above the text. If the aesthetics are found to be less pleasing - we can look at changing this (maybe not this PR).
    * matched margins and shadow style in cards and admonitions 
    * changed admonition style to look better with orange links

steffengraber reviewed (commented):
Great, a big step for the documentation. 
To improve the mobile view, I have just one small note in `custom.css`.
A general question: is it necessary to have the full bootstrap stack (js and css) in the code? Or is it not possible to leave some of it out?
One more note: There are also many content adjustments that don't necessarily have anything to do with theme customization. Wouldn't it have been better to put this in another pull request?

jessica-mitchell commented:
@steffengraber 
I adjusted the padding for smaller screens, thanks

> A general question: is it necessary to have the full bootstrap stack (js and css) in the code? 

I just followed steps from the  bootstrap site, and it might be excessive to have it all, perhaps in follow up PR I can look at simplifying this

> There are also many content adjustments that don't necessarily have anything to do with theme customization. Wouldn't it have been better to put this in another pull request?

Yah probably, but now it would be very time consuming to review commits, it largely stemmed from trying to get the output to look decent and I probably got carried away with other fixes along the way.

jessica-mitchell commented:
@heplesser I fixed up the API, including the spacing between elements in the kernel attributes which was mentioned at the NEST conference https://nest-test.readthedocs.io/en/rtd-theme/ref_material/pynest_apis.html

jessica-mitchell commented:
Instead of `models`, we can have something like `built-in models` and `custom models` and the latter could point to the NESTML page.

@pnbabu 
I was thinking about `create models` for NESTML, or maybe customize models?

jougs commented on doc/htmldoc/templates/index.html:
```diff
@@ -0,0 +1,248 @@
+    <div class="wrapper">
+      <div class="row">
+        <div class="grid_4">
+          <h1>Welcome!</h1>
+          <h2>NEST simulator documentation</h2>
+  <h3>
+  Click the button below to get NEST!
+  </h3>
+        <a href="installation/index.html" class="buttonlink">Install NEST</a>
+
+        </div>
+        <div class="grid_2">
+          <h3> Explore the docs </h3>
```
I think the H1 and H2 and H3 headlines here can go away completely. They look like some filler text without meaning and most people probably know what buttons are good for ;-)

jessica-mitchell replied:
I removed most, but I like the welcome to the documentation text, so I left that in

jougs reviewed (commented):
I just had another walk though and here's some things I noticed. I also noticed that some of the things we discussed earlier are fixed now and I like those much better now :-)

* I agree with @pnbabu and think that "Built-in models" and "Custom models" represent reality much better than just "Models" and "Customize models"  
* When I'm at the index of the [developer space](https://nest-test.readthedocs.io/en/rtd-theme/developer_space/index.html), the "Related Projects" button at the top takes me to the maze, as does "Contact us"

pnbabu commented:
> > Instead of `models`, we can have something like `built-in models` and `custom models` and the latter could point to the NESTML page.
> 
> @pnbabu I was thinking about `create models` for NESTML, or maybe `customize models`?

I see that the `Customize models` link is on the Models page. It is better to have it linked from the main page instead as `Built-in` and `Custom models`.

jessica-mitchell commented on doc/htmldoc/templates/index.html:
```diff
@@ -0,0 +1,248 @@
+         <h2 id="nest-sample"> Here is a sample NEST script. Click each section and discover related topics!</h2>
+      </div>
+
+      <div class="accordion">
```
@jougs See my comments from a few weeks ago, when I made the changes we discusssed, but basically upgrading to Bootstrap 5 seemingly caused the pulse button to get skewed and scaled improperly, and for the life of of me I coudn't get it sized and positioned correctly - I want to re-add it but at that time, I didn't want too spend too much time on one element

jessica-mitchell replied:
it works! thanks to @steffengraber

jessica-mitchell commented on doc/htmldoc/templates/index.html:
```diff
@@ -0,0 +1,248 @@
+        <a href="ref_material/pynest_apis.html" class="buttonlink">API</a>
+        </div>
+        <div class="grid_1">
+        <a href="models/index.html" class="buttonlink">Models</a>
```
Yah, I'm not super happy with the placement

jessica-mitchell replied:
adjusted the placement, I think it looks better now

jessica-mitchell commented on doc/htmldoc/templates/index.html:
```diff
@@ -0,0 +1,248 @@
+        <a href="get-started_index.html" class="buttonlink">Tutorials and examples</a>
+        </div>
+        <div class="grid_1">
+        <a href="ref_material/pynest_apis.html" class="buttonlink">API</a>
```
In a previous PR, @clinssen suggested leaving PyNEST out. The question is does the term PyNEST help users?

jessica-mitchell replied:
I still kept this as API - we can debate this at a later date?

gtrensch replied:
The terminology in the documentation should be consistent. Therefore, PyNEST API is the better choice for now.

jessica-mitchell commented on doc/htmldoc/get-started_index.rst:
```diff
@@ -3,7 +3,24 @@ Tutorials and examples
+    :gutter: 1
+
+    .. grid-item-card:: Tutorials
+       :img-top: static/img/topm2-GPS-Settings-256_nest.png 
```
icons now svg

jessica-mitchell replied:
also i figured out how to get the icon and title to be placed nicely side by side, so i opted for that because I think it looks better.

jessica-mitchell commented on doc/htmldoc/templates/index.html:
```diff
@@ -0,0 +1,248 @@
+          <img class="icon" src="_static/img/3_HBP + EBRAINS logo_Color+White.png" />
+        </div>
+        <div class="grid_4 narrow">
+          <img class="icon" src="_static/img/eu_logo.jpg" />
```
eulogo resized

gtrensch reviewed (approved):
Thank you for this extensive contribution! It looks very good to me. 👍

terhorstd commented:
This has become a very long PR with more and more changes accumulating. @jougs might have some minor points left, but @gtrensch has re-reviewed and accepted, so I'd merge this now. All lights are green.

If there are remaining points, don't hesitate to open new issues (with smaller scope).
